fw/cache.c

```diff
@@ -329,20 +330,53 @@ tfw_cache_key_node(unsigned long key)
 }
 
 /**
- * Just choose any CPU for each node to use queue_work_on() for
- * nodes scheduling. Reserve 0th CPU for other tasks.
+ * Release node-cpu map.
  */
 static void
+tfw_release_node_cpus(void)
+{
+	int node;
+
+	for(node = 0; node < nr_online_nodes; node++) {
+		if(c_nodes[node].cpu)
+			kfree(c_nodes[node].cpu);
+	}
+	kfree(c_nodes);
+}
+
+/**
+ * Create node-cpu map to use queue_work_on() for nodes scheduling.
+ * 0th CPU is reserved for other tasks.
+ * At the moment we doesn't support CPU hotplug, so enumerate only online CPUs.
+ */
+static int
 tfw_init_node_cpus(void)
 {
-	int cpu, node;
+	int nr_cpus, cpu, node;
+
+	c_nodes = kzalloc(nr_online_nodes * sizeof(CaNode), GFP_KERNEL);
+		if (!c_nodes) {
+			T_ERR( "Failed to allocate nodes map for cache work scheduler");
+			return -ENOMEM;
+	}
 
 	for_each_online_cpu(cpu) {
 		node = cpu_to_node(cpu);
+		nr_cpus = nr_cpus_node(node);
+		T_DBG("node - nr_cpus %i - %i",node, nr_cpus);
+		c_nodes[node].cpu = kmalloc(nr_cpus * sizeof(int), GFP_KERNEL);
+		if (!c_nodes[node].cpu) {
+			T_ERR( "Failed to allocate a CPU %i for cache work scheduler", cpu);
+			tfw_release_node_cpus();
```

Comment:
You should call it at the end of `tfw_cache_start` in case of error, otherwise it will never called if `tdb_open` failed for example

Reply:
This comment is not resolved! Add debug print to `tfw_release_node_cpus`! Then simulate situation when `tdb_open ` fails, you can see that `tfw_release_node_cpus` is not called! There is a memory leak. You should add `tfw_release_node_cpus` after `node_cpus_alloc_err` label  not here!